Support for Java8 interface static methods

Retrofit 2.1.0 currently throws an exception building a client implementation proxy if the interface has static methods.

static method is declared as the following in the interface:

`
static TalendClientBuilder builder() {
   return TalendClientBuilder.create();
}
`
Exception:
Caused by: java.lang.IllegalArgumentException: Could not locate call adapter for class com.corp.talend.client.v1.api.TalendClientBuilder.
  Tried:
   * retrofit2.DefaultCallAdapterFactory
	at retrofit2.Retrofit.nextCallAdapter(Retrofit.java:237) ~[retrofit-2.1.0.jar:?]
	at retrofit2.Retrofit.callAdapter(Retrofit.java:201) ~[retrofit-2.1.0.jar:?]
	at retrofit2.ServiceMethod$Builder.createCallAdapter(ServiceMethod.java:232) ~[retrofit-2.1.0.jar:?]
	at retrofit2.ServiceMethod$Builder.build(ServiceMethod.java:160) ~[retrofit-2.1.0.jar:?]
	at retrofit2.Retrofit.loadServiceMethod(Retrofit.java:166) ~[retrofit-2.1.0.jar:?]
	at retrofit2.Retrofit.eagerlyValidateMethods(Retrofit.java:156) ~[retrofit-2.1.0.jar:?]
	at retrofit2.Retrofit.create(Retrofit.java:130) ~[retrofit-2.1.0.jar:?]
...

--- a/retrofit/src/main/java/retrofit2/Retrofit.java
+++ b/retrofit/src/main/java/retrofit2/Retrofit.java
@@ -18,6 +18,7 @@ package retrofit2;
 import java.lang.annotation.Annotation;
 import java.lang.reflect.InvocationHandler;
 import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
 import java.lang.reflect.Proxy;
 import java.lang.reflect.Type;
 import java.util.ArrayList;
@@ -143,6 +144,8 @@ public final class Retrofit {
             }
             if (platform.isDefaultMethod(method)) {
               return platform.invokeDefaultMethod(method, service, proxy, args);
+            } else if (Modifier.isStatic(method.getModifiers())) {
+              return method.invoke(null, args);
             }
             return loadServiceMethod(method).invoke(args != null ? args : emptyArgs);
           }
@@ -152,7 +155,7 @@ public final class Retrofit {
   private void eagerlyValidateMethods(Class<?> service) {
     Platform platform = Platform.get();
     for (Method method : service.getDeclaredMethods()) {
-      if (!platform.isDefaultMethod(method)) {
+      if (!platform.isDefaultMethod(method) && !Modifier.isStatic(method.getModifiers())) {
         loadServiceMethod(method);
       }
     }

--- a/retrofit/src/main/java/retrofit2/Retrofit.java
+++ b/retrofit/src/main/java/retrofit2/Retrofit.java
@@ -144,8 +144,6 @@ public final class Retrofit {
             }
             if (platform.isDefaultMethod(method)) {
               return platform.invokeDefaultMethod(method, service, proxy, args);
-            } else if (Modifier.isStatic(method.getModifiers())) {
-              return method.invoke(null, args);
             }
             return loadServiceMethod(method).invoke(args != null ? args : emptyArgs);
           }

--- a/retrofit/src/test/java/retrofit2/Java8DefaultMethodsTest.java
+++ b/retrofit/src/test/java/retrofit2/Java8DefaultMethodsTest.java
@@ -25,6 +25,10 @@ public final class Java8DefaultMethodsTest {
   //  default Call<String> user() {
   //    return user("hey");
   //  }
+  //
+  //  static String staticMethod() {
+  //    return "Hi";
+  //  }
   //}
   //
   //@Test public void test() throws IOException {
@@ -34,12 +38,13 @@ public final class Java8DefaultMethodsTest {
   //  Retrofit retrofit = new Retrofit.Builder()
   //      .baseUrl(server.url("/"))
   //      .addConverterFactory(new ToStringConverterFactory())
+  //      .validateEagerly(true)
   //      .build();
   //  Example example = retrofit.create(Example.class);
   //
   //  Response<String> response = example.user().execute();
   //  assertThat(response.body()).isEqualTo("Hi");
-  //  Response<String> response = example.user("hi").execute();
-  //  assertThat(response.body()).isEqualTo("Hi");
+  //  Response<String> response2 = example.user("Hi").execute();
+  //  assertThat(response2.body()).isEqualTo("Hi");
   //}
 }

--- a/retrofit/src/test/java/retrofit2/Java8DefaultStaticMethodsInValidationTest.java
+++ b/retrofit/src/test/java/retrofit2/Java8DefaultStaticMethodsInValidationTest.java
@@ -16,7 +16,7 @@
 package retrofit2;
 
 // TODO this test doesn't play nice in the IDE because it relies on Java 8 language features.
-public final class Java8DefaultMethodsTest {
+public final class Java8DefaultStaticMethodsInValidationTest {
   //@Rule public final MockWebServer server = new MockWebServer();
   //
   //interface Example {
@@ -25,21 +25,18 @@ public final class Java8DefaultMethodsTest {
   //  default Call<String> user() {
   //    return user("hey");
   //  }
+  //
+  //  static String staticMethod() {
+  //    return "Hi";
+  //  }
   //}
   //
   //@Test public void test() throws IOException {
-  //  server.enqueue(new MockResponse().setBody("Hi"));
-  //  server.enqueue(new MockResponse().setBody("Hi"));
-  //
   //  Retrofit retrofit = new Retrofit.Builder()
   //      .baseUrl(server.url("/"))
   //      .addConverterFactory(new ToStringConverterFactory())
+  //      .validateEagerly(true)
   //      .build();
   //  Example example = retrofit.create(Example.class);
-  //
-  //  Response<String> response = example.user().execute();
-  //  assertThat(response.body()).isEqualTo("Hi");
-  //  Response<String> response2 = example.user("Hi").execute();
-  //  assertThat(response2.body()).isEqualTo("Hi");
   //}
 }

